keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,652 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+There is a corner case when Reclaim Policy is Retain and PV is referencing secrets in a certain user namespace.
+In such a case, when PVC is deleted, the PV referencing secrets remains as a resource not managed by the namespace, as a result, deletion of the secrets is blocked.
+On the other hand, provisioner secrets that are defined by using templates with PVC information,
+like "${pvc.name}", would be deleted in above case, due to the failure in finding a reference to the secret,
+because the reference to provisioner secret is not stored in PV and needs to be resolved on deletion time with PVC information that is already deleted.
+This behavior is inconsistent with other secrets, like node publish secret whose reference is stored in PV.
+
+To avoid this kind of case, an [issue](https://github.com/kubernetes-csi/external-provisioner/issues/654) for adding information on secrets in PersistentVolume as annotations is in progress. This change will allow protecting provisioner secrets from deletion even if PVC or SC for PV is deleted.
+
+## Design Details
+
+This feature can be implemented by using Lien which is being discussed in [KEP-2839](https://github.com/kubernetes/enhancements/pull/2840).
+A newly introduced Secret protection controller and a newly introduced ConfigMap protection controller update the `Liens` field of Secrets and ConfigMaps when it finds a reference to them from other resources that the controllers are watching.
+
+Note that that Secret protection controller needs to handle `VolumeSnapshotContent`, which is out-of-tree resource.
+Due to the restriction of out-of-tree resources, it needs to be handled by an external controller.
+Therefore, it is designed to make the same external controller handle both in-tree resources and out-of-tree resources, instead of creating both an in-tree controller and an external controller, separately.
+Also, ConfigMap protection controller is implemented as an external controller.
+
+To minimize the risk of unintentionally blocking the deletion, this feature is opt-in.
+Users need to add below to enable this feature per resource:
+- `kubernetes.io/enable-secret-protection` annotation to Secret
+- `kubernetes.io/enable-configmap-protection` annotation to ConfigMap
+
+Basic flows of Secret protection controller and ConfigMap protection controller are as follows:
+
+- Secret protection controller:
+  - The controller watches `Pod`, `PersistentVolume`, `VolumeSnapshotContent`, `Deployment`, `StatefulSet`, `DaemonSet`, and `CronJob`
+    - Create/Update event:
+      - For all the `Secret`s referenced by the resource:
+        1. Update the dependency graph (If the resource is using a Secret, mark the Secret that the resource is using it)
```

Comment:
Thank you for sharing the reference. I agree that similar logic will be needed and it is really helpful.
However, there will be a difference in the direction of the reference.

I understand that it gets all the [referenced secrets](https://github.com/kubernetes/kubernetes/blob/559808670ad8959008293d62cec6d252d4072616/pkg/api/v1/pod/util.go#L129) __from__ a pod, and create a list of references __from__ the pod there. On the other hand, for this use case, we need to get all the referenced resources, like pod, pvc, snapshot, __to__ the secrets and create a list of references __to__ the secret.

All the references __to__ the secret don't exist on a single resource, unlike all the references __from__ the pod to secrets exist on the single pod resource. As a result, we will need to create a list of references to a secret by checking multiple resources (And for this purpose, I think that we will need to track the referencing objects, instead of just reference count).